### PR TITLE
Allocate generation_fuel_eia923 table data to generators

### PR DESCRIPTION
**Pull request by cmgosnell**
This PR attempts to implement a solution to allocation the data from the generation_feul_eia923 table reported at the plant/fuel/prime_mover level to the plant/gen level. The bulk of this happens inside of `pudl.analysis.allocate_net_gen.py` and is integrated into `pudl.analysis.mcoe` and `pudl.output.pudltabl`

Closes Issue #777.

### CODE REVIEW

File: src/pudl/analysis/mcoe.py [hunk 1]
```diff
@@ -276,13 +281,15 @@ def capacity_factor(pudl_out, min_cap_fact=0, max_cap_fact=1.5):
...
                                           'generator_id',
                                           'capacity_mw']]
-    gen_eia923 = pudl_out.gen_eia923()[['plant_id_eia',
-                                        'report_date',
-                                        'generator_id',
-                                        'net_generation_mwh']]
+
+    gen = pudl_out.gen_eia923()
+    if pudl_out.fill_net_gen:
+        gen = pudl_out.gen_allocated()
```
Comment (rousik) [inline src/pudl/analysis/mcoe.py:287]: Would it make sense to move this conditional logic into a helper function in `pudl_out` that would check the status of `fill_net_gen` and return the corresponding dataframe?

File: src/pudl/output/pudltabl.py [hunk 2]
```diff
@@ -718,8 +731,9 @@ def hr_by_unit(self, update=False):
 
         """
         if update or self._dfs['hr_by_unit'] is None:
-            self._dfs['hr_by_unit'] = pudl.analysis.mcoe.heat_rate_by_unit(
-                self)
+            self._dfs['hr_by_unit'] = (
+                pudl.analysis.mcoe.heat_rate_by_unit(self)
```
Comment (rousik) [inline src/pudl/output/pudltabl.py:735]: Passing `self` to some other method sounds like object-oriented design gone wrong. I see that  this is not a problem introduced by this PR so I do not expect solution for it here but we should probably consider refactoring the code and eliminate this pattern in the future.

Comment (cmgosnell) [reply src/pudl/output/pudltabl.py:735]: Oh yes, the output/analysis layer needs a big overhaul. They were made originally in the very early days. I have a few ideas for this, but it'd be great to get your input on that when the time comes!

File: src/pudl/analysis/mcoe.py (hunk 1 above)
Comment (cmgosnell) [reply src/pudl/analysis/mcoe.py:287]: that is a good idea!

Comment (cmgosnell) [reply src/pudl/analysis/mcoe.py]: hey @rousik I integrated this logic into `pudl_out` as you suggested! I did a little bit of additional column management such that the columns in output of `pudl_out.gen_eia923()` is the same whether `fill_net_gen` is True or False. let me know what you think!

**Pull request merged**